### PR TITLE
Configuration Generation

### PR DESCRIPTION
**Pull request by Timidger**
If ~/.mackup.cfg does not exist, Mackup asks the user if he wants to make one. It then prompts them to answer some questions so that it can be generated. Also generates custom apps in the ~/.mackup folder if the user wants to.

### CODE REVIEW

**Comment by lra** [comment]
After giving it more thoughts, I made up my mind of having mackup to syncing anything by default. See https://github.com/lra/mackup/issues/614

But this is still super useful, except that it should be triggered by the user.

Can you make it an explicit action (e.g. `mackup configure`) instead of checking if the config file exist?

**Comment by Timidger** [comment]
Being explicit here is the better option, especially since a user might not want to answer a bunch of questions after downloading some software for the first time.  I'll get to work switching this over to be an explicit command.

**Comment by Timidger** [comment]
I switched it over to a command (`mackup configure`), and also updated the storage options to include the new ones since the patch was first written.

File: mackup/config.py [hunk 1]
```diff
@@ -25,6 +25,51 @@
...
+        (list): The list of applications that will be synced
+        (list): The list of applications that will not be synced
+    """
+    # Get the names of the apps to check the whitelist and blacklist
+    if not whitelist:
+        whitelist = []
+    if not blacklist:
+        blacklist = []
+    # Make sure the storage_type is correct
+    assert storage_type in (ENGINE_DROPBOX, ENGINE_FS,
```
Comment (lra) [inline mackup/config.py]: we have a lot more now

File: mackup/config.py [hunk 2]
```diff
@@ -25,6 +25,51 @@
...
+                            ENGINE_GDRIVE)
+    # If the path is specified, make sure it exists
+    if path:
```
Comment (lra) [inline mackup/config.py]: only allow this if the ENGINE_FS is used.

File: mackup/config.py [hunk 3]
```diff
@@ -25,6 +25,51 @@
...
+        assert os.path.exists(path), (
+            "The path {} does not exist!".format(path))
+    # Write the configuration file in the home directory
+    configuration = ["[storage]",
```
Comment (lra) [inline mackup/config.py]: Can you use Python's ConfigParser to create this file content? Way easier and solid than trying to build it with strings.

File: mackup/config.py [hunk 4]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
 
         return apps_to_sync
 
+    @staticmethod
+    def make_config_file():
+        storage_type = choose("What method would you like to use to backup "
+                              "your configuration files?",
+                              ["dropbox", "google_drive", "file_system"])[0]
```
Comment (lra) [inline mackup/config.py]: we have more

File: mackup/utils.py [hunk 5]
```diff
@@ -9,6 +9,50 @@
...
+        default(str): If given and in the options list, the default is used
+        if the user does not give an option
+
+    Returns:
+        (list): The option(s) they chose from the list
+    """
+    default_str = ("(default: {})".format(default) if default in options
+                   else "")
+    while True:
+        answer = raw_input("{} {} <{}>\n".format(question, default_str,
```
Comment (lra) [inline mackup/utils.py]: if default_str is empty, you'll have 2 spaces, which look bad.

File: mackup/config.py [hunk 6]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+
+        directory = "Mackup"
+        if confirm("Would you like to customize the directory in which "
+                         "Mackup stores your files?"):
+            directory = raw_input("Directory name: ")
+
+        whitelist = ""
+        if not confirm("Would you like to sync all applications or "
+                              "specify from a list of supported software?"):
+            whitelist = choose("Choose any of the follwing "
```
Comment (lra) [inline mackup/config.py]: following

File: mackup/config.py [hunk 7]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+            "deliminated by a space): ",
+            appsdb.ApplicationsDatabase().get_pretty_app_names(), True)
+
+        blacklist = ""
+        if confirm("Would you like to specify any applications you would "
+                        "NOT like to sync?"):
+            blacklist = choose("Choose any of the follwing "
```
Comment (lra) [inline mackup/config.py]: following

File: mackup/config.py [hunk 8]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+            "deliminated by a space): ",
+            appsdb.ApplicationsDatabase().get_pretty_app_names(), True)
+
+        custom_apps = {} # {app_name: [config files]}
+        if confirm("Now, would you like to add any applications not "
+                         "currently supported by Mackup?"):
+            print "Enter nothing to indicate your done"
```
Comment (lra) [inline mackup/config.py]: that you are done

File: mackup/config.py [hunk 9]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+            while True:
+                name = raw_input("Application name: ")
+                if not name:
+                    break
+                config_files = raw_input("Please enter the names of all the "
+                "configuration files, delemenated by a space").split()
```
Comment (lra) [inline mackup/config.py]: delimited

File: mackup/config.py [hunk 10]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+
+        custom_apps = {} # {app_name: [config files]}
+        if confirm("Now, would you like to add any applications not "
+                         "currently supported by Mackup?"):
+            print "Enter nothing to indicate your done"
+            while True:
+                name = raw_input("Application name: ")
+                if not name:
+                    break
+                config_files = raw_input("Please enter the names of all the "
```
Comment (lra) [inline mackup/config.py]: This is not gonna work, as paths can have spaces in them.

File: mackup/config.py [hunk 11]
```diff
@@ -274,6 +319,69 @@ def _parse_apps_to_sync(self):
...
+                "configuration files, delemenated by a space").split()
+                custom_apps.update({name: config_files})
+        with open(os.path.join(os.path.expanduser("~"),
+                   MACKUP_CONFIG_FILE), "w") as config_file:
+            config_file.write("[storage]\n")
```
Comment (lra) [inline mackup/config.py]: Same here, you should use Python's ConfigParser

File: mackup/utils.py [hunk 12]
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
             can_be_synced = False
 
     return can_be_synced
+
+
+def get_storage_config():
+    """
+    Asks the user to enter storage information written that will later be
```
Comment (lra) [inline mackup/utils.py]: double written

File: mackup/utils.py [hunk 13]
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
...
+            exists = os.path.exists(
+                ("" if path.startswith(os.sep) or path.startswith("./")
+                else os.environ['HOME']) + path)
+            if exists:
+                break
+            else:
+                print "The path {} does not exist!".format(exists)
+
+    while True:
+        directory = (raw_input("What would directory name would you like to "
```
Comment (lra) [inline mackup/utils.py]: Which instead of What would

File: mackup/utils.py [hunk 14]
```diff
@@ -407,3 +451,122 @@ def can_file_be_synced_on_current_platform(path):
...
+        blacklist = []
+    # Make sure the storage_type is correct
+    assert storage_type in (constants.ENGINE_DROPBOX, constants.ENGINE_FS,
+                            constants.ENGINE_GDRIVE)
+    # If the path is specified, make sure it exists
+    if path:
+        assert os.path.exists(path), (
+            "The path {} does not exist!".format(path))
+    # Write the configuration file in the home directory
+    configuration = ["[storage]",
```
Comment (lra) [inline mackup/utils.py]: Same, use ConfigParser

**Comment by Timidger** [comment]
Switched over the configuration file generation from the hacky string-based one to ConfigParser's built-in mechanism for it. I had to build it backwards though (the docs says it writes it backwards for some reason). Tested it myself as well, it generates perfectly.

**Comment by Timidger** [comment]
Updated the files. I deleted that config generator function in `utils.py`, since I moved it to `config.py` (all of its helper functions are still in `utils.py` though).

**Comment by paour** [comment]
Too bad this PR is languishing, it fixes a real pain point in mackup…

**Comment by sethherr** [comment]
:+1: I didn't know .mackup.cfg existed until looking for .gitignore_global, having documentation of it and a tool to use it (my first .mackup.cfg attempt failed because I didn't include headers) would be great.

**Comment by lra** [comment]
@sethherr PR welcome ;)

**Pull request closed without merging**